demos/dev_mode_key_provisioning/src/aws_dev_mode_key_provisioning.c

```diff
@@ -46,85 +47,500 @@
 #include "aws_clientcredential_keys.h"
 #include "aws_default_root_certificates.h"
 
-/* Key provisioning includes. */
+/* Key provisioning include. */
 #include "aws_dev_mode_key_provisioning.h"
 
+/* Utilities include. */
+#include "aws_pki_utils.h"
+
 /* mbedTLS includes. */
-#include "mbedtls/base64.h"
-/*-----------------------------------------------------------*/
+#include "mbedtls/pk.h"
+#include "mbedtls/oid.h"
 
-/* Key provisioning helper defines. */
-#define provisioningPRIVATE_KEY_TEMPLATE_COUNT         4
-#define provisioningCERTIFICATE_TEMPLATE_COUNT         3
-#define provisioningROOT_CERTIFICATE_TEMPLATE_COUNT    3
 
 /*-----------------------------------------------------------*/
 
-CK_RV xInitializePkcsSession( CK_FUNCTION_LIST_PTR * ppxFunctionList,
-                              CK_SLOT_ID * pxSlotId,
-                              CK_SESSION_HANDLE * pxSession )
+
+CK_RV xProvisionPrivateKey( CK_SESSION_HANDLE xSession,
+                            uint8_t * pucPrivateKey,
+                            size_t xPrivateKeyLength,
+                            uint8_t * pucLabel,
+                            CK_OBJECT_HANDLE_PTR pxObjectHandle )
 {
-    CK_RV xResult = 0;
-    CK_C_GetFunctionList xCkGetFunctionList = NULL;
-    CK_ULONG xCount = 1;
+    CK_RV xResult = CKR_OK;
+    CK_OBJECT_CLASS xPrivateKeyClass = CKO_PRIVATE_KEY;
+    CK_FUNCTION_LIST_PTR pxFunctionList;
+    CK_BBOOL xTokenStorage = CK_TRUE;
+    mbedtls_pk_type_t xMbedKeyType;
+    CK_KEY_TYPE xPrivateKeyType;
 
-    ( void ) ( pxSession );
+    xResult = C_GetFunctionList( &pxFunctionList );
 
-    /* Ensure that the PKCS#11 module is initialized. */
-    if( 0 == xResult )
+    int lMbedResult = 0;
+    mbedtls_pk_context xMbedPkContext;
+
+    mbedtls_pk_init( &xMbedPkContext );
+    lMbedResult = mbedtls_pk_parse_key( &xMbedPkContext, pucPrivateKey, xPrivateKeyLength, NULL, 0 );
+
+    if( lMbedResult != 0 )
     {
-        xCkGetFunctionList = C_GetFunctionList;
-        xResult = xCkGetFunctionList( ppxFunctionList );
+        xResult = CKR_ARGUMENTS_BAD;
     }
 
-    if( 0 == xResult )
+    /* Determine whether the key to be imported is RSA or EC. */
+    if( xResult == CKR_OK )
     {
-        xResult = ( *ppxFunctionList )->C_Initialize( NULL );
+        xMbedKeyType = mbedtls_pk_get_type( &xMbedPkContext );
+
+        if( xMbedKeyType == MBEDTLS_PK_RSA )
+        {
+            xPrivateKeyType = CKK_RSA;
+        }
+        else if( ( xMbedKeyType == MBEDTLS_PK_ECDSA ) || ( xMbedKeyType == MBEDTLS_PK_ECKEY ) || ( xMbedKeyType == MBEDTLS_PK_ECKEY_DH ) )
+        {
+            xPrivateKeyType = CKK_EC;
+        }
+        else
+        {
+            xPrivateKeyType = 0xFFFFFFFF;
+            xResult = CKR_ARGUMENTS_BAD;
+        }
     }
 
-    /* Get the default private key storage ID. */
-    if( ( 0 == xResult ) || ( CKR_CRYPTOKI_ALREADY_INITIALIZED == xResult ) )
+    /* EC Keys. */
+    if( xPrivateKeyType == CKK_EC )
```

Comment:
Is it ok to keep going to this if if the xResult form the last block is not CKR_OK?